### ISSUE TITLE
LSF: specific requirements for submitting jobs on our cluster

### ISSUE THREAD

**Issue by aslett1** [marbl/canu#13]
Hi,

We use LSF as our grid engine (I realise that support for LSF is untested, of course). Our systems team have set a requirement, when setting memory limits that, in addition to setting "-M MEMORY", we also need to set a couple of -R options (-R 'select [mem>MEMORY] rusage [mem=MEMORY]') so, for example, if a job requests 6144MB of memory , we need to set the options:

-M 6144 -R 'select[mem>6144] rusage [mem=6144]'

If the -R option is not present, or, if the memory values for -R differ from that set with -M, a pre-check on all LSF submissions causes the job to fail (actually, it isn't submitted).

What changes would we need to make to include this in our checkout of canu? I'd, maybe naively, assumed that just changing the line "setGlobalIfUndef("gridEngineMemoryOption",               "-M MEMORY");" in Grid_LSF.pm might be the only requirement but, unless I'm missing something, other changes are required?

We can obviously set 'useGrid=0' and then bsub the job, thus running outside of the grid set up of canu nut still under LSF, but this seems like an inefficient use of resources, outside of testing things. (I did this earlier today and it worked once I'd tweaked the required number of threads)

Any help would be much appreciated!

Thanks in advance,

Martin

**Comment by brianwalenz** [comment]
That should be the only change - just add the -R '...' to the end of the string.  Or, pass this in through gridEngineMemoryOption="-M MEMORY -R '...'" on the canu command line (or specfile) and leave the code unmodified.

**Comment by aslett1** [comment]
Thanks! I'm giving the gridEngineMemoryOption a go. Got an error first 
time up which may have been down to threaded jobs running on multiple 
hosts (adding span [hosts=1] to the -R option).

As an aside, the link to the gzipped 25X subset FASTQ E. coli reads from 
the readthedoces page throws a 404 error (using curl to get the unzipped 
file works fine).

Best regards,

Martin

On Tue, 5 Jan 2016, brianwalenz wrote:

> 
> That should be the only change - just add the -R '...' to the end of the string. Or, pass this in through
> gridEngineMemoryOption="-M MEMORY -R '...'" on the canu command line (or specfile) and leave the code unmodified.
> 
> —
> Reply to this email directly or view it on GitHub.[AAwjN4XE4XmX0u6oaTSQe7MjDOXQQbn7ks5pW9aYgaJpZM4G-4xi.gif]
> 
> 
>


-- 
 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

**Comment by aslett1** [comment]
Hi again,

Setting gridEngineMemoryOption on the command line is working but my jobs are still failing. This time with a non-existent job array after the first job runs. It looks like the gridEngineHoldOption setting is causing problems. I've tried changing this on the command line but I'm not sure exactly what the format needs to be. I think the issue is that numended requires the job ID and the -w check is using job name?

I've attached the output file.

Thanks,

Martin


[canu.01.out.txt](https://github.com/marbl/canu/files/79917/canu.01.out.txt)

**Comment by skoren** [comment]
Yep, now you're in the untested code based on LSF documentation since we don't have a system to test on. The LSF documentation claims that done, ended, etc take either job name or job id but it might be the case that numended doesn't follow this behavior. You can try modifying 
src/pipelines/canu/Execution.pm, line 803 from:

```
        if (uc(getGlobal("gridEngine")) eq "PBS" || uc(getGlobal("gridEngine")) eq "SLURM"){
```

to

```
        if (uc(getGlobal("gridEngine")) eq "PBS" || uc(getGlobal("gridEngine")) eq "SLURM" || uc(getGlobal("gridEngine") eq "LSF"){
```
and see if that makes the holds valid. However, there would still be untested LSF code in terms of job restarts/resumes.

**Comment by txje** [comment]
This fix worked for me: https://github.com/marbl/canu/pull/39

**Comment by brianwalenz** [comment]
And that was just merged in commit 7b16f2ffdef4041fe2cc5c25629c353efb1de2ea.  Thanks!
